### PR TITLE
[FEATURE] main dev of the service

### PR DESCRIPTION
**Pull request by LitalFiverr**

### CODE REVIEW

File: api/feature_api.rb [hunk 1]
```diff
@@ -1,25 +1,138 @@
...
+      feature_name = params[:name]
+      error!('Resource not found.', 404) if feature_name.nil? || !Feature.exist?(feature_name)
+    end
+
+    def current_feature
+      feature_name = params[:name]
+      return nil if feature_name.nil?
+
+      Feature.find(feature_name)
+    end
```
Comment (LitalFiverr) [inline api/feature_api.rb:14]: how about using params coercions for both validation and getting the feature ? (Never used it but looks suitable)
https://github.com/ruby-grape/grape#custom-types-and-coercions
(you can add the `params do` block for `name` just above the namespace wrap with an empty namespace  the methods where it is needed, that will have the `params do` block).

File: api/feature_api.rb [hunk 2]
```diff
@@ -1,25 +1,138 @@
...
+    RestfulModels::Response.represent(data: features)
   end
 
-  post '/:feature' do
-    feature_name = params[:feature]
-    percentage = params[:percentage] || 0
-    $rollout.activate_percentage(feature_name, percentage)
-    $rollout.set_feature_data(feature_name,  params[:meta_data])
-    {}
+  route_param :name do
```
Comment (LitalFiverr) [inline api/feature_api.rb:27]: it is more of a `cached_slug` then a `name`. :)

File: api/feature_api.rb [hunk 3]
```diff
@@ -1,25 +1,138 @@
...
+      feature_exist!
+      feature = current_feature
+
+      feature.delete
+      RestfulModels::Response.represent(message: 'The feature has been removed.')
+    end
+
+    params do
+      requires :description, type: String, desc: 'The feature description'
+      requires :author, type: String, desc: 'The author name'
```
Comment (LitalFiverr) [inline api/feature_api.rb:91]: do we want name ? maybe email/slack name ?

File: api/feature_api.rb [hunk 4]
```diff
@@ -1,25 +1,138 @@
...
+          created_at: Time.current,
+          created_by: params[:author]
+      }
+
+      feature = Feature.new(options)
+
+      begin
+        feature.save!
+        RestfulModels::Response.represent(message: 'Feature created successfully!')
+      rescue Exception
```
Comment (LitalFiverr) [inline api/feature_api.rb:108]: i think `rescue => e` will catch all exceptions that you want to catch. 
http://daniel.fone.net.nz/blog/2013/05/28/why-you-should-never-rescue-exception-in-ruby/
(how about sharing the exception on the message ? we wouldn't like to share it?)

File: config/white_list_ips.yml [hunk 5]
```diff
@@ -0,0 +1,7 @@
+development:
+  - 127.0.0.1
+production:
+  -
+  -
+  -
+  -
```
Comment (LitalFiverr) [inline config/white_list_ips.yml:7]: remove this as we discussed

File: models/feature.rb [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+    feature_data.merge!({
+     name: feature.name,
+     percentage: feature.percentage,
+     users: feature.users
+    })
+    self.new(feature_data)
+  end
+
+  def self.exist?(name)
+    @features ||= $rollout.features
```
Comment (LitalFiverr) [inline models/feature.rb:50]: this will be initalized every request, not sure you need the whole list in order to check if one item exists. 
can you just check if the feature exists on the set with the gem ?

File: models/feature.rb [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+
+  def add_user(user_id)
+    $rollout.activate_user(self.name, user_id)
+  end
+
+  def remove_user(user_id)
+    $rollout.deactivate_user(self.name, user_id)
+  end
+
+  def set_default_values
```
Comment (LitalFiverr) [inline models/feature.rb:89]: can be private

File: models/feature.rb [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+    self.history ||= []
+    self.users ||= []
+    self.percentage ||= 0
+  end
+
+  def set_history_attribute
```
Comment (LitalFiverr) [inline models/feature.rb:95]: can be private

File: models/feature.rb [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+
+    $rollout.activate_percentage(self.name, self.percentage)
+
+    feature_data = {
+        history: self.history,
+        description: self.description,
+        author: self.author,
+        created_at: self.created_at,
+        created_by: self.created_by,
+    }
```
Comment (LitalFiverr) [inline models/feature.rb:66]: what if the feature already had data? is it possible ? in this case it will be overriden, are we OK with that? :fire:

File: syslib/globals.rb [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+  def environment
+    $env = ENV['RACK_ENV'] || 'development'
+  end
+
+  def redis
+    config =  {
+        host: ENV['redis_host'] || '127.0.0.1',
+        port: ENV['redis_port'] || 6379
+    }
+    $redis = Redis.new(config)
```
Comment (LitalFiverr) [inline syslib/globals.rb:13]: I think the configuration should be on yml. 
This case is the same as chimera service uses active-record gem. 
the definitions for the active-record sits in databases.yml. 
The different is that weirdly  rollout gem gets a redis/sotrage instance instead of configuration it deals with.
The best things was if the rollout.gem hasd a configuration yml as mongoid/active-record has. 
I guess they didn't get to it yet :)

File: syslib/globals.rb [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+  end
+
+  def rollout
+    $rollout = Rollout.new($redis)
+  end
+
+  def white_list_ips
+    $white_list_ips = YAML.load(File.open('./config/white_list_ips.yml'))[$env] || []
+  end
```
Comment (LitalFiverr) [inline syslib/globals.rb:22]: can be removed

**Review by LitalFiverr — COMMENTED** [review]
very nice :) clean and readable code  that uses best practices!

File: restful_models/response.rb [hunk 12]
```diff
@@ -0,0 +1,6 @@
+module RestfulModels
+   class Response < Grape::Entity
+      expose :data, unless: Proc.new {|field| field.nil?}
+      expose :message,unless: Proc.new {|field| field.nil?}
```
Comment (roeepolegfiverr) [inline restful_models/response.rb:4]: [style] missing space. make sure to init and run Rubocop

File: models/feature.rb [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+                :created_at,
+                :created_by
+
+  def initialize(attributes)
+    super(attributes)
+    self.set_default_values
+  end
+
+  def self.find(name)
+    return nil unless exist?(name)
```
Comment (roeepolegfiverr) [inline models/feature.rb:35]: why is that needed? just try to fetch it and check the result.

File: models/feature.rb [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+
+    feature = $rollout.get(name)
+    feature_data = feature.data.symbolize_keys!
+    feature_data.delete_if {|key, _| !self.method_defined?(key)}
+
+    feature_data.merge!({
+     name: feature.name,
```
Comment (roeepolegfiverr) [inline models/feature.rb:42]: i don't see any change here (same field names), why do you create a new hash and merge it back?

File: models/feature.rb [hunk 15]
```diff
@@ -0,0 +1,121 @@
...
+        author: self.author,
+        created_at: self.created_at,
+        created_by: self.created_by,
+    }
+
+    feature_data.delete_if { |_, value| value.blank? }
+    $rollout.set_feature_data(self.name, feature_data)
+  end
+
+  def active?(user_id)
```
Comment (roeepolegfiverr) [inline models/feature.rb:72]: `active_for_user_id?` ?

File: models/feature.rb [hunk 16]
```diff
@@ -0,0 +1,121 @@
...
+    $rollout.active?(self.name, user_id)
+  end
+
+  def delete
+    $rollout.clear_feature_data(self.name)
```
Comment (roeepolegfiverr) [inline models/feature.rb:77]: just delete

File: models/feature.rb [hunk 17]
```diff
@@ -0,0 +1,121 @@
...
+  def self.exist?(name)
+    @features ||= $rollout.features
+    @features.include?(name.to_sym)
+  end
+
+  def save!
+    self.set_history_attribute
+    Throw 'Feature is not valid!' unless self.valid?
+
+    $rollout.activate_percentage(self.name, self.percentage)
```
Comment (roeepolegfiverr) [inline models/feature.rb:58]: how come save has nothing to do with users? 
Regardless of the API of Rollout i think this method should handle that too.

File: models/feature.rb [hunk 18]
```diff
@@ -0,0 +1,121 @@
+require 'active_model'
```
Comment (roeepolegfiverr) [inline models/feature.rb:1]: I suggest using `active_attr` gem for Restfull models. 
You'r getting validations, default values, nice mongoid-like class syntax and more.

File: syslib/globals.rb [hunk 19]
```diff
@@ -0,0 +1,32 @@
...
+  def redis
+    config =  {
+        host: ENV['redis_host'] || '127.0.0.1',
+        port: ENV['redis_port'] || 6379
+    }
+    $redis = Redis.new(config)
+  end
+
+  def rollout
+    $rollout = Rollout.new($redis)
```
Comment (roeepolegfiverr) [inline syslib/globals.rb:17]: use `$rollout = Rollout.new($redis, use_sets: true)`

File: api/feature_api.rb (hunk 1 above)
Comment (yossi-eynav) [reply api/feature_api.rb:14]: Great comment!

File: api/feature_api.rb (hunk 2 above)
Comment (yossi-eynav) [reply api/feature_api.rb:27]: `feature_name` ?

File: api/feature_api.rb (hunk 3 above)
Comment (yossi-eynav) [reply api/feature_api.rb:91]: what about both?  email and name?
Phase 2 will be to authenticate the users

File: api/feature_api.rb (hunk 4 above)
Comment (yossi-eynav) [reply api/feature_api.rb:108]: Great!

File: models/feature.rb (hunk 6 above)
Comment (yossi-eynav) [reply models/feature.rb:50]: you can query the feature (from rollout gem) but incase it doesn't exist, it creates a new one(?!).
That's why i'm doing it myself

File: models/feature.rb (hunk 9 above)
Comment (yossi-eynav) [reply models/feature.rb:66]: Yes, why should we care?
you're gonna be thrown out  in the API level if the feature exists and you're trying to recreate it (instead of update)

File: models/feature.rb (hunk 13 above)
Comment (yossi-eynav) [reply models/feature.rb:35]: what do you mean?

File: models/feature.rb (hunk 14 above)
Comment (yossi-eynav) [reply models/feature.rb:42]: `feature_data` contains the meta data of the feature
`feature` is an object created by rollout.

I merging the 2 so I can create a new instance of the Feature model

File: models/feature.rb (hunk 15 above)
Comment (yossi-eynav) [reply models/feature.rb:72]: could be also for a group/ feature open  so the name `active_for_user_id ` is misleading

File: syslib/globals.rb (hunk 10 above)
Comment (yossi-eynav) [reply syslib/globals.rb:13]: Your'e expecting too much from a gem that returns "OK" as the response of it's actions 🙂

File: models/feature.rb (hunk 18 above)
Comment (yossi-eynav) [reply models/feature.rb:1]: AWESOME :deploy

File: api/feature_api.rb [hunk 20]
```diff
@@ -1,25 +1,138 @@
...
-    $rollout.set_feature_data(feature_name,  params[:meta_data])
-    {}
+  route_param :name do
+
+    params do
+      requires :user_id, type: Integer, desc: 'The user ID'
+    end
+    route_param :user_id do
+      post '/' do
+        feature_exist!
```
Comment (ramigluzman) [inline api/feature_api.rb:34]: Use `before` blocks instead of calling to `feature_exist!` for every route

File: api/feature_api.rb (hunk 3 above)
Comment (ramigluzman) [reply api/feature_api.rb:91]: I suggest to use the HTTP credentials as we do in impersonator -https://github.com/fiverr/impersonator/blob/master/impersonator.rb#L38

File: api/feature_api.rb [hunk 21]
```diff
@@ -1,25 +1,138 @@
...
+    end
+
+    params do
+      requires :description, type: String, desc: 'The feature description'
+      requires :author, type: String, desc: 'The author name'
+    end
+    post '/' do
+      options = {
+          name: params[:name],
+          percentage: params[:percentage] || 0,
```
Comment (ramigluzman) [inline api/feature_api.rb:96]: ```params[:percentage].to_i```

File: api/feature_api.rb (hunk 4 above)
Comment (ramigluzman) [reply api/feature_api.rb:108]: The more-standard way (as we do in other chimeras) is to handle grape exceptions using ```rescue_all` - this will save you the rescue block for each route.
Example: https://github.com/fiverr/mobile_chimera/blob/master/service/mobile_chimera.rb#L24

File: api/feature_api.rb [hunk 22]
```diff
@@ -1,25 +1,138 @@
...
+
+      feature.assign_attributes(options)
+
+      begin
+        feature.save!
+        RestfulModels::Response.represent(message: 'Feature updated successfully!')
+      rescue Exception
+        status 500
+        RestfulModels::Response.represent(message: 'An error has been accord')
+      end
```
Comment (ramigluzman) [inline api/feature_api.rb:135]: See my comment above ⏫

File: config.ru [hunk 23]
```diff
@@ -1,16 +1,13 @@
 require 'bundler'
 Bundler.require
 
+require_rel 'syslib'
+Globals.setup
+
+# require_rel 'models'
+require_rel 'restful_models'
```
Comment (ramigluzman) [inline config.ru:8]: You can use ```require_all``` gem

File: api/feature_api.rb (hunk 3 above)
Comment (yossi-eynav) [reply api/feature_api.rb:91]: @ramigluzman 
phase 2 will be to implement google oauth / github oauth

File: api/feature_api.rb (hunk 4 above)
Comment (yossi-eynav) [reply api/feature_api.rb:108]: @ramigluzman 
I'll add it in addition - as a failsafe

File: api/feature_api.rb (hunk 20 above)
Comment (yossi-eynav) [reply api/feature_api.rb:34]: Removed the helpers
